### PR TITLE
Make ampersand optional for some common blocks

### PR DESCRIPTION
**Pull request by STRd6**

### CODE REVIEW

File: source/parser.hera [hunk 1]
```diff
@@ -1323,7 +1323,7 @@ FunctionExpression
     }
 
   # Ruby/Crystal style block shorthand
-  AmpersandUnaryPrefix?:prefix Ampersand AmpersandBlockRHS?:rhs ->
+  AmpersandUnaryPrefix?:prefix ( Ampersand / ( !NumericLiteral &( QuestionMark? Dot ) ) ) AmpersandBlockRHS?:rhs ->
```
Comment (edemaine) [inline source/parser.hera:1326]: What does the `!NumericLiteral` do?

File: test/call-expression.civet [hunk 2]
```diff
@@ -132,11 +132,15 @@ describe "call-expression", ->
...
-    it "should not parse with !. on newline after plain identifier", ->
-      throws """
-        data
-        !.map (x) => x + 1
-      """
+    testCase """
+      !. on newline after plain identifier is a separate fn block
+      ---
+      data
+      !.map (x) => x + 1
```
Comment (edemaine) [inline test/call-expression.civet:139]: Took me a while to understand what this does... Unary operators with implicit `&` is a little crazy. But I guess it makes sense. (Though in this case doesn't make sense to negate an array :))

File: test/function-block-shorthand.civet [hunk 3]
```diff
@@ -17,6 +17,22 @@ describe "&. function block shorthand", ->
...
+    x.map($ => $.name)
+  """
+
+  testCase """
+    omit &
+    ---
+    x.map ?.name
+    ---
+    x.map($ => $?.name)
+  """
```
Comment (edemaine) [inline test/function-block-shorthand.civet:34]: These are definitely cool!

**Review by edemaine — APPROVED** [review]

File: source/parser.hera (hunk 1 above)
Comment (STRd6) [reply source/parser.hera:1326]: It prevents `.1` from being interpreted as `&.1`.

File: test/call-expression.civet (hunk 2 above)
Comment (STRd6) [reply test/call-expression.civet:139]: Yeah, it's a little strange but falls out naturally from the existing rules.

File: source/parser.hera (hunk 1 above)
Comment (edemaine) [reply source/parser.hera:1327]: Does it? Wouldn't it have to come after the Dot, inside the positive lookahead assertation?

Comment (edemaine) [reply source/parser.hera:1327]: Oh, I see, I was thinking of `/[0-9]/`, but `NumericLiteral` includes a leading dot. Got it!

**Pull request merged**